jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/handler/ConditionalHandlerTest.java

```diff
@@ -0,0 +1,260 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.eclipse.jetty.http.pathmap.PathSpec;
+import org.eclipse.jetty.server.ForwardedRequestCustomizer;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.LocalConnector;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.not;
+
+public class ConditionalHandlerTest
+{
+    private Server _server;
+    private LocalConnector _connector;
+    private HelloHandler _helloHandler;
+
+    @BeforeEach
+    public void beforeEach() throws Exception
+    {
+        _server = new Server();
+        _connector = new LocalConnector(_server);
+        _connector.getConnectionFactory(HttpConnectionFactory.class).getHttpConfiguration().addCustomizer(new ForwardedRequestCustomizer());
+        _server.addConnector(_connector);
+        _helloHandler = new HelloHandler();
+    }
+
+    @AfterEach
+    public void afterEach() throws Exception
+    {
+        _server.stop();
+    }
+
+    private void startServer(Handler.Singleton testHandler) throws Exception
+    {
+        _server.setHandler(testHandler);
+        testHandler.getTail().setHandler(_helloHandler);
+        _server.start();
+    }
+
+    public static Stream<ConditionalHandler> conditionalHandlers()
+    {
+        return Stream.of(
```

Comment:
NB if you change the sense of the Dont/Skip/ForbidHandlers to do their thang in `doHandle()` instead of  `doNotHandle()` then you will probably have to change the sense of these tests.